### PR TITLE
Add AWS provider

### PR DESCRIPTION
**Pull request by hhamalai**
Add autoscaler provider to execute builds on AWS

### CODE REVIEW

**Comment by anbraten** [comment]
Thanks for your work. I guess AWS is affected by https://github.com/woodpecker-ci/autoscaler/issues/91 as well, right?

**Comment by henkka** [comment]
> Thanks for your work. I guess AWS is affected by #91 as well, right?

I believe this pull request is not affected by the issue mentioned in #91 because the AWS Instance Metadata Service v2 hop limit configuration in https://github.com/woodpecker-ci/autoscaler/pull/118/files#diff-c77d8b5077c4d2c3350ad7ca994ceffe1737935b4e1390f0963acca2d338ca67R74 prevents the containers from connecting to the Instance Metadata API 🤔

https://www.element7.io/2023/01/shift-left-security-why-you-should-use-aws-imdsv2-explained-in-detail/ seems to summarize the hop limit as

> It becomes interesting once you know that with IMDSv2, the PUT response containing the secret token can, by default, not travel outside the instance. This is accomplished by having a default Time To Live (TTL) on the low-level IP packets containing the secret token set to “1”. Hardware and software that handle packets, including EC2 instances, subtract 1 from each packet’s TTL whenever they pass it on. If the TTL gets to 0, the packet is discarded. A packet with a TTL of “n” can therefore make n “hops” in a network before giving up, while a packet with a TTL of “1” can exist in just one. In the context of IMDSv2, we refer to hop limit of the PUT call instead of TTL. By using a default hop limit (TTL) of 1, a session token can only be used directly from the EC2 instance where that session was initiated.

and without the IMDS session token the cloud-init user-data cannot be fetched

File: providers/aws/provider.go [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+			{
+				Name:   aws.String("tag:Name"),
+				Values: []string{agent.Name},
+			},
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+	return &instances.Reservations[0].Instances[0], nil
```
Comment (anbraten) [inline providers/aws/provider.go:140]: Might be good idea to check again out-of-bounds here?!

File: providers/aws/flags.go [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+	},
+	&cli.StringFlag{
+		Name:     "aws-ami-id",
+		Usage:    "aws ami id",
+		EnvVars:  []string{"WOODPECKER_AWS_AMI_ID"},
+		Category: Category,
+	},
+	&cli.StringSliceFlag{
+		Name:     "aws-tags",
+		Usage:    "aws tags",
```
Comment (anbraten) [inline providers/aws/flags.go:23]:
```suggestion
		Usage:    "additional tags for your ec2 instances",
```

File: providers/aws/flags.go [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+	},
+	&cli.StringFlag{
+		Name:     "aws-region",
+		Usage:    "aws region",
+		EnvVars:  []string{"WOODPECKER_AWS_REGION"},
+		Category: Category,
+	},
+	&cli.StringSliceFlag{
+		Name:     "aws-subnets",
+		Usage:    "aws subnets",
```
Comment (anbraten) [inline providers/aws/flags.go:35]: Not 100% getting what value I would set here. Could you adjust the usage / description value and add an example to it.

File: providers/aws/provider.go [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+		return err
+	}
+	_, err = p.client.TerminateInstances(ctx, &ec2.TerminateInstancesInput{
+		InstanceIds: []string{*instance.InstanceId},
+	})
+	return err
+}
+
+func (p *Provider) ListDeployedAgentNames(ctx context.Context) ([]string, error) {
+	zerolog.Debug().Msgf("List deployed agent name")
```
Comment (anbraten) [inline providers/aws/provider.go:155]:
```suggestion
	zerolog.Debug().Msgf("List deployed agent names")
```

**Review by anbraten — COMMENTED** [review]
lgtm. some minor things

**Review by anbraten — APPROVED** [review]

**Review by 6543 — APPROVED** [review]

**Pull request merged**